homeassistant/components/media_player/kef.py

```diff
@@ -0,0 +1,302 @@
+"""
+Support for interfacing HA with the KEF Wireless Speakers .
+
+For more details about this platform, please refer to the documentation at
+https://github.com/Gronis/pykef
+"""
+
+
+import collections
+import logging
+import time
+import json
+from enum import Enum
+import voluptuous as vol
+
+from homeassistant.components.media_player import (
+    PLATFORM_SCHEMA, SUPPORT_SELECT_SOURCE, SUPPORT_VOLUME_MUTE,
+    SUPPORT_VOLUME_SET, SUPPORT_TURN_ON, SUPPORT_VOLUME_STEP, SUPPORT_TURN_OFF,
+    MediaPlayerDevice
+)
+from homeassistant.const import (
+    CONF_HOST, CONF_NAME, CONF_PORT, STATE_OFF, STATE_ON
+)
+from homeassistant.helpers import config_validation as cv
+
+REQUIREMENTS = ['pykef==1.2.0']
+
+_CONFIGURING = {}
+
+_LOGGER = logging.getLogger(__name__)
+
+
+DEFAULT_NAME = 'KEFWirelessSpeaker'
+DEFAULT_PORT = 50001
+DATA_KEF = 'kef'
+# If a new source is selected, do not override source in update for this amount
+#  of seconds
+UPDATE_TIMEOUT = 1.0
+# When turning off/on the speaker, do not query it for CHANGE_STATE_TIMEOUT,
+# since it takes
+# some time for it to go offline/onlne
+CHANGE_STATE_TIMEOUT = 30.0
+# If we try to control the speaker while offline, wait for the speaker to come
+# online (in secs)
+WAIT_FOR_ONLINE_STATE = 10.0
+
+# supported features
+SUPPORT_KEF_CLIENT_DEVICE = (
+    SUPPORT_VOLUME_SET | SUPPORT_VOLUME_STEP | SUPPORT_VOLUME_MUTE |
+    SUPPORT_SELECT_SOURCE | SUPPORT_TURN_OFF
+)
+
+CONF_TURN_ON_SERVICE = 'turn_on_service'
+CONF_TURN_ON_DATA = 'turn_on_data'
+
+# yaml configuration
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_HOST): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_TURN_ON_SERVICE): cv.service,
+    vol.Optional(CONF_TURN_ON_DATA): cv.string,
+})
+
+
+# setup of component
+def setup_platform(hass, config, add_devices,
+                   discovery_info=None):
+    """Set up Kef platform."""
+    import pykef
+    host = config.get(CONF_HOST)
+    port = config.get(CONF_PORT)
+    name = config.get(CONF_NAME)
+    speaker = pykef.KefSpeaker(host, port)
+    turn_on_service = config.get(CONF_TURN_ON_SERVICE)
+    turn_on_data = config.get(CONF_TURN_ON_DATA)
+
+    # configure source options to communicate to HA
+    sources = collections.OrderedDict([
+        ('1', str(pykef.InputSource.Wifi)),
+        ('2', str(pykef.InputSource.Bluetooth)),
+        ('3', str(pykef.InputSource.Aux)),
+        ('4', str(pykef.InputSource.Opt)),
+        ('5', str(pykef.InputSource.Usb))
+    ])
+
+    _LOGGER.debug(
+        "Setting up %s, using host: %s, port: %s, name: %s",
+        DATA_KEF, str(host), str(port), str(name))
+    _LOGGER.debug("Setting up source_dict %s", str(sources))
+
+    # Add devices
+    add_devices([KefClientDevice(
+        name, speaker, turn_on_service, turn_on_data, sources, hass
```

Comment:
Update the entity -> https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#5-entity

Reply:
Ack